server/src/utils/DumpPdfUtils.ts

```diff
@@ -0,0 +1,236 @@
+          if (pageKids && hasNestedKeys(pageKids, ['list', 'ref'])) {
+            pagesInfo.push(...pageKids.list.ref.map(r => r.$.id));
+          }
+        }
+      }
+
+      if (nodeIs(node, 'Type', 'Page')) {
+        foundNodeType = true;
+        if (hasNestedKeys(node, ['dict', 'key']) && hasNestedKeys(node, ['dict', 'value'])) {
+          const annotPage = node.dict.value[node.dict.key.findIndex(k => k === 'Annots')];
+          if (!pagesWithAnnots[node.$.id]) {
+            pagesWithAnnots[node.$.id] = [];
+          }
+          if (annotPage && hasNestedKeys(annotPage, ['list', 'ref'])) {
+            pagesWithAnnots[node.$.id].push(...annotPage.list.ref.map(r => r.$.id));
+          } else if (annotPage && hasNestedKeys(annotPage, ['ref', '$', 'id'])) {
+            pagesWithAnnots[node.$.id].push(annotPage.ref.$.id);
+            nodesToDeepSearch.push(annotPage.ref.$.id);
+          }
+        }
+      }
+
+      if (nodeIs(node, 'Type', 'Annot') || nodeIs(node, 'Subtype', 'Link')) {
+        foundNodeType = true;
+        if (hasNestedKeys(node, ['dict', 'key']) && hasNestedKeys(node, ['dict', 'value'])) {
+          const a = node.dict.value[node.dict.key.findIndex(k => k === 'A')];
+          if (hasNestedKeys(a, ['ref', '$', 'id']) && !hasNestedKeys(a, ['dict'])) {
+            nodesToDeepSearch.push(a.ref.$.id);
+          } else {
+            annots.push(node);
+          }
+        }
+      }
+
+      // 'Action' type nodes have targetURLs, but don't include bbox data, so they are ignored.
+      if (nodeIs(node, 'Type', 'Action')) {
+        foundNodeType = true;
+      }
+
+      if (
+        nodesToDeepSearch.includes(node.$.id)
+        && !foundNodeType
+        && !hasNestedKeys(node, ['dict', 'value'])
+      ) {
+        linkNodes.push(node);
+      }
+    });
+
+    xmlStream.on('error', message => {
+      reject(message);
+    });
+
+    xmlStream.on('end', () => {
+      pagesInfo = pagesInfo.filter(p => Object.keys(pagesWithAnnots).includes(p));
+      const links: DumpPdfLink[] = annots.map(xmlNodeToLink).filter(l => !!l);
+      const linksData = Object.keys(pagesWithAnnots)
+        .map(nodeId => {
+          return {
+            pageNumber: pagesInfo.findIndex(p => p === nodeId),
+            links: pagesWithAnnots[nodeId]
+              .map(matchLinks(links, linkNodes.map(xmlToLinkingNode)))
+              .reduce((acc, array) => acc.concat(...array), []),
+          };
+        })
+        .filter(p => p.links.length > 0);
+
+      resolve(linksData);
+    });
+  });
+}
+
+/*
+  this looks that a particular nested structure defined in @param keys exists on @param o.
+  ex:
+    hasNestedKeys({ foo: { bar: { baz: 10 } } }, ['foo', 'bar', 'baz']) === true, structure "o.foo.bar.baz" is valid
+    hasNestedKeys({ foo: { bar: { baz: 10 } } }, ['foo', 'baz']) === false, structure "o.foo.baz" is not valid
+*/
+function hasNestedKeys(o: any, keys: string[]): boolean {
```

Comment:
`o` should be node or element or whatever that makes easy to understand function signature ??

Reply:
solved in https://github.com/axa-group/Parsr/pull/376/commits/ab0f5da1437acaa78cd562112c2f6ca207196e7e